## Exercise for Chapter 3

This exercises is design to assist you how to use the pandas package to import, preprocess data and perform basic statistical analysis. Later we should see how data generating events can produce data of interest to insurance analysts.

We will look at the Local Government Property Insurance Fund in this chapter. The fund insures property owned by municipal governments, such as schools and libraries.

* government buildings,

* educational institutions,

* public libraries, and

* motor vehicles.

Over a thousand local government units are covered by the fund, which charges about \$25 million in annual premiums and provides insurance coverage of about \$75 billion.

**Example 1** Import the claim dataset namely ClaimsExperienceData.csv from my Github repository. Then write Python commands to answer the following questions.

In [1]:
import pandas as pd

claims = pd.read_csv('/Users/Kaemyuijang/SCMA248/Data/ClaimsExperienceData.csv')

1. How many claims observations are there in this dataset?

In [2]:
claims.shape[0]

5639

2. How many variables (features) are there in this dataset? List (print out) all the features. 

In [3]:
claims.shape[1]
claims.columns

Index(['PolicyNum', 'Year', 'LnCoverage', 'BCcov', 'Premium', 'Freq', 'Deduct',
       'y', 'lny', 'yAvg', 'lnDeduct', 'Fire5', 'NoClaimCredit', 'TypeCity',
       'TypeCounty', 'TypeMisc', 'TypeSchool', 'TypeTown', 'TypeVillage',
       'AC00', 'AC05', 'AC10', 'AC15'],
      dtype='object')


## Description of Rating Variables

One of the important tasks of insurance analysts is to develop models to represent and manage the two outcome variables, **frequency** and **severity**. 

However, when actuaries and other financial analysts use those models, they do so in the context of external variables. 

In general statistical terminology, one might call these explanatory or predictor variables.

Because of our insurance focus, we call them **rating variables** as they are useful in setting insurance rates and premiums.

The following table describes the rating variables considered.

These are variables that you think might naturally be related to claims outcomes.

<!-- To handle the skewness, we henceforth focus on logarithmic transformations of coverage and deductibles. -->

<!-- For our immediate purposes, the coverage is our first rating variable. Other things being equal, we would expect that policyholders with larger coverage have larger claims. We will make this vague idea much more precise as we proceed, and also justify this expectation with data. -->

**Variable**  | **Description**
----- | -------------
EntityType    | Categorical variable that is one of six types: (Village, City, County, Misc, School, or Town) 
LnCoverage    | Total building and content coverage, in logarithmic millions of dollars
LnDeduct      | Deductible, in logarithmic dollars
AlarmCredit   | Categorical variable that is one of four types: (0, 5, 10, or 15) for automatic smoke alarms in main rooms
NoClaimCredit | Binary variable to indicate no claims in the past two years
Fire5         | Binary variable to indicate the fire class is below 5 (The range of fire class is 0 to 10)  

**In what follows, for illustrate, we will consider claims data in year 2010.**

3. How many policies are there in 2010? 

Name the answer with the variable name **num_policies**. 

Hint: one may use `.value_counts` method that return a Series containing counts of unique values. Alternatively, you want to count False and True separately you can use `pd.Series.sum()` + `~`.

In [4]:
temp = claims['Year']  == 2010
temp.value_counts()

False    4529
True     1110
Name: Year, dtype: int64

In [5]:
num_policies = temp.sum()

In [6]:
(~temp).sum()

4529

4. How many claims are there in 2010? Assign the result to the variable **num_claims**.

In [7]:
claims2010 = claims[temp]

In [8]:
claims2010.columns

Index(['PolicyNum', 'Year', 'LnCoverage', 'BCcov', 'Premium', 'Freq', 'Deduct',
       'y', 'lny', 'yAvg', 'lnDeduct', 'Fire5', 'NoClaimCredit', 'TypeCity',
       'TypeCounty', 'TypeMisc', 'TypeSchool', 'TypeTown', 'TypeVillage',
       'AC00', 'AC05', 'AC10', 'AC15'],
      dtype='object')

In [9]:
claims2010.sum()

PolicyNum        1.652057e+08
Year             2.231100e+06
LnCoverage       2.488480e+03
BCcov            4.577870e+10
Premium          1.590532e+07
Freq             1.377000e+03
Deduct           3.994500e+06
y                3.665931e+07
lny              3.736787e+03
yAvg             2.270177e+07
lnDeduct         8.013576e+03
Fire5            6.220000e+02
NoClaimCredit    6.270000e+02
TypeCity         1.560000e+02
TypeCounty       7.100000e+01
TypeMisc         1.230000e+02
TypeSchool       3.110000e+02
TypeTown         1.850000e+02
TypeVillage      2.640000e+02
AC00             3.460000e+02
AC05             8.200000e+01
AC10             9.000000e+01
AC15             5.920000e+02
dtype: float64

In [10]:
num_claims = claims2010['Freq'].sum()
print(num_claims)

1377


5. Which policy number has the maximum number of claims and what is this claims number?

In [11]:
claims2010.sort_values('Freq', ascending = False).head(2)

PolicyNum  Year  LnCoverage         BCcov  Premium  Freq   Deduct  \
1406     138109  2010    6.338472  5.659308e+08   124504   239  25000.0   
100      120030  2010    7.801717  2.444797e+09   391168   103  50000.0   

               y        lny          yAvg  ...  TypeCity  TypeCounty  \
1406   223784.65  12.318439    936.337448  ...         0           0   
100   4920530.65  15.408927  47772.142230  ...         0           1   

      TypeMisc  TypeSchool  TypeTown  TypeVillage  AC00  AC05  AC10  AC15  
1406         0           1         0            0     0     0     0     1  
100          0           0         0            0     0     0     0     1  

[2 rows x 23 columns]

# Hard cording

claims2010.loc[1406,'Freq'] 

With `.idxmax()`, we can return the index at which maximum weight value is present.

See https://www.geeksforgeeks.org/get-the-index-of-maximum-value-in-dataframe-column/.

In [12]:
print(claims2010['Freq'].idxmax())

ind_freq_max = claims2010['Freq'].idxmax()

max_claims = claims2010.loc[ind_freq_max,'Freq'] 

1406


6. Calculate the proportion of policyholders who did not have any claims (use the name **num_policies_no_claims** for your output).

In [13]:
# Using value_count() and .sort_index to obtain the number of 
# policies by claim numbers.

(claims2010['Freq'].value_counts()).sort_index()

num_policies_no_claims = (claims2010['Freq'].value_counts()).sort_index()[0]

In [14]:
# Calculate the proportion of policyholders who did not have any claims.

round(num_policies_no_claims / num_policies,4)

0.6369

In [15]:
(claims2010['Freq'].value_counts())[0]/claims2010['Freq'].sum()

0.5134350036310821

7. Calculate the proportion of policyholders who had only one claim.

In [16]:
num_policies_one_claims = (claims2010['Freq'].value_counts()).sort_index()[1]

In [17]:
round(num_policies_one_claims / num_policies,4)

0.1883

8. Calculate the average number of claims for this sample. 

In [18]:
num_claims/num_policies

1.2405405405405405

9. The `describe()` method is used for calculating some statistical data like percentile, mean and std of the numerical values of the Series or DataFrame. 

Applying to year 2010, what do we get when we run the command claims.describe()?

In [19]:
claims2010.describe()

PolicyNum    Year   LnCoverage         BCcov        Premium  \
count    1110.000000  1110.0  1110.000000  1.110000e+03    1110.000000   
mean   148833.995495  2010.0     2.241874  4.124207e+07   14329.113514   
std     16131.790893     0.0     1.962844  1.142432e+08   24663.572338   
min    120002.000000  2010.0    -4.575223  1.030400e+04       9.000000   
25%    138104.250000  2010.0     0.900512  2.460876e+06    1633.500000   
50%    150302.000000  2010.0     2.523816  1.247615e+07    6365.000000   
75%    160628.750000  2010.0     3.705105  4.065431e+07   17923.750000   
max    180791.000000  2010.0     7.801717  2.444797e+09  391168.000000   

              Freq         Deduct             y          lny          yAvg  \
count  1110.000000    1110.000000  1.110000e+03  1110.000000  1.110000e+03   
mean      1.240541    3598.648649  3.302640e+04     3.366475  2.045205e+04   
std       8.154437    8787.925562  4.287782e+05     4.573141  3.927241e+05   
min       0.000000     500.000000  0.000000e+00     0.000000  0.000000e+00   
25%       0.000000     500.000000  0.000000e+00     0.000000  0.000000e+00   
50%       0.000000    1000.000000  0.000000e+00     0.000000  0.000000e+00   
75%       1.000000    2500.000000  4.139750e+03     8.328310  2.818750e+03   
max     239.000000  100000.000000  1.292222e+07    16.374459  1.292222e+07   

       ...     TypeCity   TypeCounty     TypeMisc  TypeSchool     TypeTown  \
count  ...  1110.000000  1110.000000  1110.000000  1110.00000  1110.000000   
mean   ...     0.140541     0.063964     0.110811     0.28018     0.166667   
std    ...     0.347704     0.244799     0.314039     0.44929     0.372846   
min    ...     0.000000     0.000000     0.000000     0.00000     0.000000   
25%    ...     0.000000     0.000000     0.000000     0.00000     0.000000   
50%    ...     0.000000     0.000000     0.000000     0.00000     0.000000   
75%    ...     0.000000     0.000000     0.000000     1.00000     0.000000   
max    ...     1.000000     1.000000     1.000000     1.00000     1.000000   

       TypeVillage         AC00         AC05         AC10         AC15  
count  1110.000000  1110.000000  1110.000000  1110.000000  1110.000000  
mean      0.237838     0.311712     0.073874     0.081081     0.533333  
std       0.425951     0.463402     0.261683     0.273083     0.499113  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     1.000000  
75%       0.000000     1.000000     0.000000     0.000000     1.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 23 columns]

10. A common method for determining the severity distribution is to look at the distribution of the sample of 1,377 claims. Another typical strategy is to look at the **distribution of average claims among policyholders who have made claims**.

In our 2010 sample, how many such policyholders who have made claims?

In [20]:
num_policies - num_policies_no_claims

403

11. The average claim for the 209 policyholders who had only one claim is the same as the single claim they had. 

Write the command(s) to list the average claim of such 209 policyholders.

In [21]:
selected_index = (claims2010['Freq'] == 1)

claims2010[selected_index][['Freq','y']]


Freq          y
4        1    6838.87
9        1    9711.28
14       1   10323.50
24       1    3469.79
31       1   35000.00
...    ...        ...
5534     1    1851.48
5568     1    3405.00
5603     1   20679.58
5635     1  168304.05
5636     1    1034.33

[209 rows x 2 columns]

12. Calculate the average claim of the policyholder with the maximum number of claims.

ind_freq_max = claims2010['Freq'].idxmax()

max_claims = claims2010.loc[ind_freq_max,'Freq'] 

In [22]:
claims2010.loc[ind_freq_max,'y'] / claims2010.loc[ind_freq_max,'Freq'] 

936.3374476987448

In [23]:
claims.describe()

PolicyNum         Year   LnCoverage         BCcov        Premium  \
count    5639.000000  5639.000000  5639.000000  5.639000e+03    5639.000000   
mean   148880.491222  2007.979784     2.132881  3.728085e+07   14796.029793   
std     15911.165649     1.415949     1.977179  1.034020e+08   25520.718048   
min    120002.000000  2006.000000    -4.717555  8.937000e+03       9.000000   
25%    138112.000000  2007.000000     0.785066  2.192556e+06    1612.500000   
50%    150350.000000  2008.000000     2.429532  1.135357e+07    6578.000000   
75%    160619.000000  2009.000000     3.606218  3.682654e+07   18184.000000   
max    180791.000000  2010.000000     7.801717  2.444797e+09  412328.000000   

              Freq         Deduct             y          lny          yAvg  \
count  5639.000000    5639.000000  5.639000e+03  5639.000000  5.639000e+03   
mean      1.109239    3364.869658  1.728730e+04     2.753088  9.291565e+03   
std       8.549179    8273.670512  2.304627e+05     4.311270  1.975141e+05   
min       0.000000     500.000000  0.000000e+00     0.000000  0.000000e+00   
25%       0.000000     500.000000  0.000000e+00     0.000000  0.000000e+00   
50%       0.000000    1000.000000  0.000000e+00     0.000000  0.000000e+00   
75%       1.000000    2500.000000  2.181825e+03     7.687917  1.519706e+03   
max     263.000000  100000.000000  1.292222e+07    16.374459  1.292222e+07   

       ...     TypeCity   TypeCounty     TypeMisc   TypeSchool     TypeTown  \
count  ...  5639.000000  5639.000000  5639.000000  5639.000000  5639.000000   
mean   ...     0.140628     0.058166     0.107998     0.283206     0.172194   
std    ...     0.347668     0.234078     0.310405     0.450596     0.377582   
min    ...     0.000000     0.000000     0.000000     0.000000     0.000000   
25%    ...     0.000000     0.000000     0.000000     0.000000     0.000000   
50%    ...     0.000000     0.000000     0.000000     0.000000     0.000000   
75%    ...     0.000000     0.000000     0.000000     1.000000     0.000000   
max    ...     1.000000     1.000000     1.000000     1.000000     1.000000   

       TypeVillage         AC00         AC05         AC10         AC15  
count  5639.000000  5639.000000  5639.000000  5639.000000  5639.000000  
mean      0.237808     0.463025     0.042383     0.057989     0.436602  
std       0.425779     0.498675     0.201480     0.233743     0.496008  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     1.000000     0.000000     0.000000     1.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 23 columns]

In [24]:
claims.mean()

PolicyNum        1.488805e+05
Year             2.007980e+03
LnCoverage       2.132881e+00
BCcov            3.728085e+07
Premium          1.479603e+04
Freq             1.109239e+00
Deduct           3.364870e+03
y                1.728730e+04
lny              2.753088e+00
yAvg             9.291565e+03
lnDeduct         7.157236e+00
Fire5            5.552403e-01
NoClaimCredit    3.286044e-01
TypeCity         1.406278e-01
TypeCounty       5.816634e-02
TypeMisc         1.079979e-01
TypeSchool       2.832062e-01
TypeTown         1.721937e-01
TypeVillage      2.378081e-01
AC00             4.630254e-01
AC05             4.238340e-02
AC10             5.798901e-02
AC15             4.366022e-01
dtype: float64

## Part 2

1. Write Python code to generage a table that shows the 2010 claims frequency distribution. The table should contain the number of policies, the number of claims and the proportion (broken down by the number of claims).

Goal: the table should tell us how many poicyholders and the (percentage) proportion of policyholders who did not have any claims, only one claim and so on.

1.1. How many policyholders in the 2010 claims data have 9 or more claims?

1.2. What is the percentage proportion of policyholders having exactly 3 claims?

In [25]:
import pandas as pd

#claims = pd.read_csv('/Users/Kaemyuijang/SCMA248/Data/ClaimsExperienceData.csv')
url = 'https://raw.githubusercontent.com/pairote-sat/SCMA248/main/Data/ClaimsExperienceData.csv'
claims = pd.read_csv(url)

URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1091)>

To obtain the number of policies and number of claims categorized by claim frequency (and by year), we first group `claims` dataset by `Year` and `Freq` and apply `agg` (aggregate) function.

In [3]:
claimsFreqDist = claims.groupby(['Year','Freq']).agg({'PolicyNum' : 'count', 'Freq' : 'sum'})
claimsFreqDist.rename(columns={'PolicyNum':'NumPolicies', 'Freq':'NumClaims'}, inplace = True)
claimsFreqDist

NumPolicies  NumClaims
Year Freq                        
2006 0             845          0
     1             153        153
     2              67        134
     3              40        120
     4              13         52
...                ...        ...
2010 19              1         19
     30              1         30
     39              1         39
     103             1        103
     239             1        239

[103 rows x 2 columns]

To determine the (percentage) proportion of policies for each claim frequency, we will first group the "Year" and "Freq".   To calculate the percentage **within each "Year" group**, the following command can be used groupby(level=0).apply(lambda x: 100*x/x.sum())

**Note:** Because the original dataframe becomes a multiple index dataframe after grouping, the level = 0 refers to the top level index, which in our case is 'Year'.

You can see the results below, which have already been sorted by percent of policies for each claim frequency by year.

In [12]:
#Table1 = claims.groupby(['Year','Freq']).agg({'PolicyNum' : 'count'}).groupby(level='Year').apply(lambda x: 100*x/x.sum())
Table1 = claims.groupby(['Year','Freq']).agg({'PolicyNum' : 'count'}).groupby(level='Year').apply(lambda x: 100*x/x.sum())
Table1.rename(columns={'PolicyNum':'Percentage'}, inplace = True)

#Table1.loc[(2010,slice(None))]
Table1.loc[(slice(None),slice(None))].tail(30)

Percentage
Year Freq            
2009 16      0.089928
     18      0.089928
     19      0.089928
     21      0.089928
     143     0.089928
     228     0.089928
     263     0.089928
2010 0      63.693694
     1      18.828829
     2       7.747748
     3       3.603604
     4       1.621622
     5       1.081081
     6       0.810811
     7       0.360360
     8       0.540541
     9       0.090090
     10      0.270270
     11      0.180180
     13      0.090090
     14      0.180180
     15      0.090090
     16      0.180180
     17      0.090090
     18      0.090090
     19      0.090090
     30      0.090090
     39      0.090090
     103     0.090090
     239     0.090090

The final is to merged the above two resulting tables using `pd.merge` function.

In [13]:
claimsFreqDist = pd.merge(Table1,claimsFreqDist, how = 'left', on = ['Year','Freq'])

#claimsFreqDist

claimsFreqDist.loc[(2010,slice(None))].sort_index(axis=1)

NumClaims  NumPolicies  Percentage
Freq                                    
0             0          707   63.693694
1           209          209   18.828829
2           172           86    7.747748
3           120           40    3.603604
4            72           18    1.621622
5            60           12    1.081081
6            54            9    0.810811
7            28            4    0.360360
8            48            6    0.540541
9             9            1    0.090090
10           30            3    0.270270
11           22            2    0.180180
13           13            1    0.090090
14           28            2    0.180180
15           15            1    0.090090
16           32            2    0.180180
17           17            1    0.090090
18           18            1    0.090090
19           19            1    0.090090
30           30            1    0.090090
39           39            1    0.090090
103         103            1    0.090090
239         239            1    0.090090

In [318]:
#claimsFreqDist.loc[(2010,slice(None))].sort_index(axis=1)

claimsFreqDist.index
claimsFreqDist.loc[(slice(None),[1,2,5]),['Percentage','NumClaims'] ].head()

Percentage  NumClaims
Year Freq                       
2006 1      13.258232        153
     2       5.805893        134
     5       0.779896         45
2007 1      12.829525        146
     2       8.084359        184

1.1. How many policyholders in the 2010 claims data have 9 or more claims?

In [58]:
# see for more detail: https://stackoverflow.com/questions/50608749/slicing-a-multiindex-dataframe-with-a-condition-based-on-the-index

#l0 and l1 are Year and Freq levels, respectively.
l0 = claimsFreqDist.index.get_level_values(0)
l1 = claimsFreqDist.index.get_level_values(1)
cond = (l0 == 2010) & (l1 >= 9)

claimsFreqDist.loc[cond,'NumPolicies'].sum()

ans1_1 = claimsFreqDist.loc[cond,'NumPolicies'].sum()

print('Ans: The number of policyholders in the 2010 claims data having 9 or more claims is'
      , ans1_1)

Ans: The number of policyholders in the 2010 claims data having 9 or more claims is 19


In [309]:
# https://stackoverflow.com/questions/53927460/select-rows-in-pandas-multiindex-dataframe

claimsFreqDist.query('Year == 2010 & Freq >= 9')['NumPolicies'].sum()

19

1.2. What is the percentage proportion of policyholders having exactly 3 claims?

In [59]:
claimsFreqDist.query('Year == 2010 & Freq == 3')['Percentage']

ans1_2 = (claimsFreqDist.query('Year == 2010 & Freq == 3')['Percentage']).values[0]


print('Ans: The percentage proportion of policyholders having exactly 3 claims:', ans1_2)

Ans: The percentage proportion of policyholders having exactly 3 claims: 3.6036036036036037


In [43]:
claims2010 = claims[claims['Year']  == 2010]

In [44]:
claims2010.head()

PolicyNum  Year  LnCoverage        BCcov  Premium  Freq  Deduct         y  \
4      120002  2010    3.157489   23511493.0     7994     1  1000.0   6838.87   
9      120003  2010    4.741850  114646079.0    36687     1  5000.0   9711.28   
14     120004  2010    3.461150   31853574.0    17839     1  1000.0  10323.50   
19     120005  2010    3.487794   32713698.0    12431     0  5000.0      0.00   
24     120008  2010    3.856969   47321724.0    16090     1   500.0   3469.79   

         lny      yAvg  ...  TypeCity  TypeCounty  TypeMisc  TypeSchool  \
4   8.830378   6838.87  ...         0           1         0           0   
9   9.181043   9711.28  ...         0           1         0           0   
14  9.242178  10323.50  ...         0           1         0           0   
19  0.000000      0.00  ...         0           1         0           0   
24  8.151849   3469.79  ...         0           1         0           0   

    TypeTown  TypeVillage  AC00  AC05  AC10  AC15  
4          0            0     1     0     0     0  
9          0            0     0     0     0     1  
14         0            0     0     0     0     1  
19         0            0     0     0     0     1  
24         0            0     0     0     0     1  

[5 rows x 23 columns]

In [49]:
claims.columns

claims.index.names

FrozenList([None])

This outputs a “FrozenList”, which is just a Pandas specific construct used to show the index label(s) of a DataFrame. Here, we see the value is “None”, as this is the default value of a DataFrame’s index.

To create a MultiIndex with our original DataFrame, all we need to do is pass a list of columns into the .set_index() Pandas function like this:

In [52]:
multi = claims.set_index(['Year','Freq'])

multi.tail(50)

PolicyNum  LnCoverage        BCcov  Premium    Deduct          y  \
Year Freq                                                                     
2006 0        180775   -0.088613     915200.0      943     500.0       0.00   
2007 0        180775   -0.021734     978500.0      753     500.0       0.00   
2008 0        180775    0.011596    1011664.0      769     500.0       0.00   
2009 0        180775    0.049213    1050444.0      799     500.0       0.00   
2010 0        180775    0.022723    1022983.0      717     500.0       0.00   
2006 0        180776   -2.571773      76400.0      101     500.0       0.00   
2007 0        180776   -2.511616      81137.0       72     500.0       0.00   
2008 0        180776   -2.451468      86167.0       68     500.0       0.00   
2009 0        180776   -2.436574      87460.0       69     500.0       0.00   
2010 0        180776   -2.420704      88859.0       64     500.0       0.00   
2006 0        180777    0.973993    2648500.0     3152    1000.0       0.00   
2007 0        180777    1.119558    3063500.0     1807    1000.0       0.00   
2008 0        180777    2.114208    8283020.0     5384    1000.0       0.00   
2009 0        180777    2.157580    8650181.0     5277    1000.0       0.00   
2010 1        180777    2.145404    8545496.0     5554    1000.0   20679.58   
2006 0        180778    1.715362    5558687.0     5448     500.0       0.00   
2007 0        180778    1.407991    4087735.0     2903     500.0       0.00   
2008 0        180778    2.047009    7744705.0     3872    5000.0       0.00   
2009 0        180778    2.110863    8255364.0     3219    5000.0       0.00   
2010 0        180778    2.082594    8025258.0     3210    5000.0       0.00   
2007 1        180779   -1.107149     330500.0      427     500.0    1708.00   
2008 0        180779    0.901121    2462363.0     2118     500.0       0.00   
2009 0        180779    0.933579    2543596.0     2035     500.0       0.00   
2010 0        180779    0.839066    2314204.0     1829     500.0       0.00   
2007 0        180780    2.768237   15930517.0     8921    5000.0       0.00   
2008 0        180780    2.768237   15930517.0    11629    5000.0       0.00   
     0        180781   -1.386294     250000.0      196     500.0       0.00   
2009 0        180781   -1.363864     255671.0      199     500.0       0.00   
2010 0        180781   -1.316026     268199.0      187     500.0       0.00   
2008 0        180782   -0.517565     595970.0      429     500.0       0.00   
2009 0        180782   -0.504327     603912.0      326     500.0       0.00   
2010 0        180782   -0.542029     581567.0      337     500.0       0.00   
2008 0        180783   -1.722629     178596.0      191    1000.0       0.00   
2009 1        180783   -1.684199     185593.0      143    1000.0   11905.40   
2010 0        180783   -1.713239     180281.0      147    1000.0       0.00   
2008 0        180784   -0.693147     500000.0      160    1000.0       0.00   
2009 0        180784   -0.088831     915000.0      320    1000.0       0.00   
2010 0        180784   -0.049608     951602.0      362    1000.0       0.00   
2008 0        180785    0.000000    1000000.0     1120    1000.0       0.00   
2009 0        180785    0.893987    2444859.0     1076    1000.0       0.00   
2010 0        180785    0.844678    2327228.0     1094    1000.0       0.00   
2009 0        180786    1.963733    7125879.0     9192     500.0       0.00   
2010 0        180786    1.999797    7387558.0     5245     500.0       0.00   
2008 0        180787   -1.609438     200000.0      200    1000.0       0.00   
2009 0        180787   -2.001133     135182.0       92     500.0       0.00   
2010 0        180787   -1.255266     285000.0      199     500.0       0.00   
     1        180788    6.032462  416739800.0    58344  100000.0  168304.05   
     1        180789   -0.691173     500988.0      295     500.0    1034.33   
     0        180790    1.275549    3580665.0     2077    100

In [6]:
multi.index.names

FrozenList(['Year', 'Freq'])

In [7]:
multi.index.values

array([(2006, 0), (2007, 0), (2008, 0), ..., (2010, 1), (2010, 0),
       (2010, 0)], dtype=object)

In [9]:
multi = claims.set_index(['Year','Freq']).sort_index()

multi.head()

PolicyNum  LnCoverage        BCcov  Premium   Deduct    y  lny  \
Year Freq                                                                   
2006 0        120002    3.123002   22714456.0     9313   1000.0  0.0  0.0   
     0        120003    4.598683   99353377.0    40735   5000.0  0.0  0.0   
     0        120005    3.448799   31462584.0    19192   1000.0  0.0  0.0   
     0        120009    4.198984   66618639.0    15988  25000.0  0.0  0.0   
     0        120018    4.621174  101613238.0    51823   5000.0  0.0  0.0   

           yAvg   lnDeduct  Fire5  ...  TypeCity  TypeCounty  TypeMisc  \
Year Freq                          ...                                   
2006 0      0.0   6.907755      1  ...         0           1         0   
     0      0.0   8.517193      1  ...         0           1         0   
     0      0.0   6.907755      1  ...         0           1         0   
     0      0.0  10.126631      1  ...         0           1         0   
     0      0.0   8.517193      1  ...         0           1         0   

           TypeSchool  TypeTown  TypeVillage  AC00  AC05  AC10  AC15  
Year Freq                                                             
2006 0              0         0            0     1     0     0     0  
     0              0         0            0     0     0     0     1  
     0              0         0            0     0     0     0     1  
     0              0         0            0     0     0     0     1  
     0              0         0            0     0     0     0     1  

[5 rows x 21 columns]

In [11]:
multi.reset_index()

Year  Freq  PolicyNum  LnCoverage         BCcov  Premium   Deduct  \
0     2006     0     120002    3.123002  2.271446e+07     9313   1000.0   
1     2006     0     120003    4.598683  9.935338e+07    40735   5000.0   
2     2006     0     120005    3.448799  3.146258e+07    19192   1000.0   
3     2006     0     120009    4.198984  6.661864e+07    15988  25000.0   
4     2006     0     120018    4.621174  1.016132e+08    51823   5000.0   
...    ...   ...        ...         ...           ...      ...      ...   
5634  2010    19     120012    6.344524  5.693661e+08   222052    500.0   
5635  2010    30     140337    6.336490  5.648104e+08   129907  25000.0   
5636  2010    39     120073    5.492783  2.429324e+08    34010  50000.0   
5637  2010   103     120030    7.801717  2.444797e+09   391168  50000.0   
5638  2010   239     138109    6.338472  5.659308e+08   124504  25000.0   

               y        lny          yAvg  ...  TypeCity  TypeCounty  \
0           0.00   0.000000      0.000000  ...         0           1   
1           0.00   0.000000      0.000000  ...         0           1   
2           0.00   0.000000      0.000000  ...         0           1   
3           0.00   0.000000      0.000000  ...         0           1   
4           0.00   0.000000      0.000000  ...         0           1   
...          ...        ...           ...  ...       ...         ...   
5634    83934.70  11.337794   4417.615789  ...         0           1   
5635   759020.97  13.539785  25300.699000  ...         1           0   
5636    69943.80  11.155447   1793.430769  ...         0           1   
5637  4920530.65  15.408927  47772.142230  ...         0           1   
5638   223784.65  12.318439    936.337448  ...         0           0   

      TypeMisc  TypeSchool  TypeTown  TypeVillage  AC00  AC05  AC10  AC15  
0            0           0         0            0     1     0     0     0  
1            0           0         0            0     0     0     0     1  
2            0           0         0            0     0     0     0     1  
3            0           0         0            0     0     0     0     1  
4            0           0         0            0     0     0     0     1  
...        ...         ...       ...          ...   ...   ...   ...   ...  
5634         0           0         0            0     0     0     0     1  
5635         0           0         0            0     0     0     0     1  
5636         0           0         0            0     0     0     0     1  
5637         0           0         0            0     0     0     0     1  
5638         0           1         0            0     0     0     0     1  

[5639 rows x 23 columns]

In [22]:
multi.loc[(2010,0),['y','Deduct']]

y   Deduct
Year Freq              
2010 0     0.0   5000.0
     0     0.0  25000.0
     0     0.0    500.0
     0     0.0   5000.0
     0     0.0   5000.0
...        ...      ...
     0     0.0   1000.0
     0     0.0    500.0
     0     0.0    500.0
     0     0.0   1000.0
     0     0.0    500.0

[707 rows x 2 columns]

In [25]:
multi.loc[([2008,2010],[0,1]),['y','Deduct']]

y    Deduct
Year Freq                     
2008 0          0.00    1000.0
     0          0.00     500.0
     0          0.00    5000.0
     0          0.00   25000.0
     0          0.00    2500.0
...              ...       ...
2010 1       1851.48     500.0
     1       3405.00    1000.0
     1      20679.58    1000.0
     1     168304.05  100000.0
     1       1034.33     500.0

[1887 rows x 2 columns]

In [5]:
out = claims.groupby(by = ['Year','Freq']).count()
out.head()

PolicyNum  LnCoverage  BCcov  Premium  Deduct    y  lny  yAvg  \
Year Freq                                                                  
2006 0           845         845    845      845     845  845  845   845   
     1           153         153    153      153     153  153  153   153   
     2            67          67     67       67      67   67   67    67   
     3            40          40     40       40      40   40   40    40   
     4            13          13     13       13      13   13   13    13   

           lnDeduct  Fire5  ...  TypeCity  TypeCounty  TypeMisc  TypeSchool  \
Year Freq                   ...                                               
2006 0          845    845  ...       845         845       845         845   
     1          153    153  ...       153         153       153         153   
     2           67     67  ...        67          67        67          67   
     3           40     40  ...        40          40        40          40   
     4           13     13  ...        13          13        13          13   

           TypeTown  TypeVillage  AC00  AC05  AC10  AC15  
Year Freq                                                 
2006 0          845          845   845   845   845   845  
     1          153          153   153   153   153   153  
     2           67           67    67    67    67    67  
     3           40           40    40    40    40    40  
     4           13           13    13    13    13    13  

[5 rows x 21 columns]

In [31]:
out.index

MultiIndex([(2006,   0),
            (2006,   1),
            (2006,   2),
            (2006,   3),
            (2006,   4),
            (2006,   5),
            (2006,   6),
            (2006,   7),
            (2006,   8),
            (2006,   9),
            ...
            (2010,  14),
            (2010,  15),
            (2010,  16),
            (2010,  17),
            (2010,  18),
            (2010,  19),
            (2010,  30),
            (2010,  39),
            (2010, 103),
            (2010, 239)],
           names=['Year', 'Freq'], length=103)

In [63]:
out.loc[[2006,], 'y']

Year  Freq
2006  0       845
      1       153
      2        67
      3        40
      4        13
      5         9
      6         6
      7         1
      8         6
      9         3
      10        3
      11        2
      14        1
      20        1
      24        1
      58        1
      100       1
      208       1
Name: y, dtype: int64

In [68]:
out.loc[[2010,], 'y']

Year  Freq
2010  0       707
      1       209
      2        86
      3        40
      4        18
      5        12
      6         9
      7         4
      8         6
      9         1
      10        3
      11        2
      13        1
      14        2
      15        1
      16        2
      17        1
      18        1
      19        1
      30        1
      39        1
      103       1
      239       1
Name: y, dtype: int64

In [69]:
out.index.names

FrozenList(['Year', 'Freq'])

Start here!!!

Again, we pass a tuple in with our desired index values, but instead of adding values for “freq”, we pass `slice(None)`. This is the default `slice` command in Pandas to select all the contents of the MultiIndex level.

In [52]:
out2010 = out.loc[(2010,slice(None)),:]

In [45]:
out2010.columns

Index(['PolicyNum', 'LnCoverage', 'BCcov', 'Premium', 'Deduct', 'y', 'lny',
       'yAvg', 'lnDeduct', 'Fire5', 'NoClaimCredit', 'TypeCity', 'TypeCounty',
       'TypeMisc', 'TypeSchool', 'TypeTown', 'TypeVillage', 'AC00', 'AC05',
       'AC10', 'AC15'],
      dtype='object')

In [46]:
100*out2010['PolicyNum']/(out2010['PolicyNum'].sum())

Year  Freq
2010  0       63.693694
      1       18.828829
      2        7.747748
      3        3.603604
      4        1.621622
      5        1.081081
      6        0.810811
      7        0.360360
      8        0.540541
      9        0.090090
      10       0.270270
      11       0.180180
      13       0.090090
      14       0.180180
      15       0.090090
      16       0.180180
      17       0.090090
      18       0.090090
      19       0.090090
      30       0.090090
      39       0.090090
      103      0.090090
      239      0.090090
Name: PolicyNum, dtype: float64

In [47]:
out2010.shape

(23, 21)

In [28]:
out2010.index

MultiIndex([(2010,   0),
            (2010,   1),
            (2010,   2),
            (2010,   3),
            (2010,   4),
            (2010,   5),
            (2010,   6),
            (2010,   7),
            (2010,   8),
            (2010,   9),
            (2010,  10),
            (2010,  11),
            (2010,  13),
            (2010,  14),
            (2010,  15),
            (2010,  16),
            (2010,  17),
            (2010,  18),
            (2010,  19),
            (2010,  30),
            (2010,  39),
            (2010, 103),
            (2010, 239)],
           names=['Year', 'Freq'])

In [42]:
(out2010.loc[(2010,slice(None)),:])['Proportion']

Year  Freq
2010  0       63.693694
      1       18.828829
      2        7.747748
      3        3.603604
      4        1.621622
      5        1.081081
      6        0.810811
      7        0.360360
      8        0.540541
      9        0.090090
      10       0.270270
      11       0.180180
      13       0.090090
      14       0.180180
      15       0.090090
      16       0.180180
      17       0.090090
      18       0.090090
      19       0.090090
      30       0.090090
      39       0.090090
      103      0.090090
      239      0.090090
Name: Proportion, dtype: float64

In [25]:
(100*out2010['PolicyNum']/(out2010['PolicyNum'].sum())).shape

(23,)

In [48]:
out2010['Proportion'] = 100*out2010['PolicyNum']/(out2010['PolicyNum'].sum())

/Users/Kaemyuijang/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [51]:
out2010.loc[:,['PolicyNum','Proportion']]

PolicyNum  Proportion
Year Freq                       
2010 0           707   63.693694
     1           209   18.828829
     2            86    7.747748
     3            40    3.603604
     4            18    1.621622
     5            12    1.081081
     6             9    0.810811
     7             4    0.360360
     8             6    0.540541
     9             1    0.090090
     10            3    0.270270
     11            2    0.180180
     13            1    0.090090
     14            2    0.180180
     15            1    0.090090
     16            2    0.180180
     17            1    0.090090
     18            1    0.090090
     19            1    0.090090
     30            1    0.090090
     39            1    0.090090
     103           1    0.090090
     239           1    0.090090

In [53]:
out2010 = out.loc[(2010,slice(None)),:]

out2010.index


MultiIndex([(2010,   0),
            (2010,   1),
            (2010,   2),
            (2010,   3),
            (2010,   4),
            (2010,   5),
            (2010,   6),
            (2010,   7),
            (2010,   8),
            (2010,   9),
            (2010,  10),
            (2010,  11),
            (2010,  13),
            (2010,  14),
            (2010,  15),
            (2010,  16),
            (2010,  17),
            (2010,  18),
            (2010,  19),
            (2010,  30),
            (2010,  39),
            (2010, 103),
            (2010, 239)],
           names=['Year', 'Freq'])

In [65]:
# This works pretty simply, but the resulting DataFrames no longer have the multi-index. Also .xs() is not the most powerful way to subset a DataFrame.
# https://www.somebits.com/~nelson/pandas-multiindex-slice-demo.html


out2010 = (out.xs(2010))

print(out2010.index)
out2010['Proportion'] = 100*out2010['PolicyNum']/(out2010['PolicyNum'].sum())

out2010.loc[:,['PolicyNum','Proportion']]

Int64Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  13,
             14,  15,  16,  17,  18,  19,  30,  39, 103, 239],
           dtype='int64', name='Freq')


PolicyNum  Proportion
Freq                       
0           707   63.693694
1           209   18.828829
2            86    7.747748
3            40    3.603604
4            18    1.621622
5            12    1.081081
6             9    0.810811
7             4    0.360360
8             6    0.540541
9             1    0.090090
10            3    0.270270
11            2    0.180180
13            1    0.090090
14            2    0.180180
15            1    0.090090
16            2    0.180180
17            1    0.090090
18            1    0.090090
19            1    0.090090
30            1    0.090090
39            1    0.090090
103           1    0.090090
239           1    0.090090

## Next attempt

In [106]:
table1 = claims.loc[:,['Year','Freq','PolicyNum']].groupby(by = ['Year','Freq']).count()
#table1['PolicyNum']

#print(table1.loc[(2006,),:])

#print(table1.head())

table1.groupby('Year').sum()

table1.loc[(2006,slice(None))]



AttributeError: 'DataFrameGroupBy' object has no attribute 'names'

## Pivot

In [26]:
print(claims.columns)
func = lambda x: 100*x.count()/1110
claims.pivot_table(index = 'Freq', columns = 'Year', values = ['PolicyNum','PolicyNum'], aggfunc = ['count','sum'])

Index(['PolicyNum', 'Year', 'LnCoverage', 'BCcov', 'Premium', 'Freq', 'Deduct',
       'y', 'lny', 'yAvg', 'lnDeduct', 'Fire5', 'NoClaimCredit', 'TypeCity',
       'TypeCounty', 'TypeMisc', 'TypeSchool', 'TypeTown', 'TypeVillage',
       'AC00', 'AC05', 'AC10', 'AC15'],
      dtype='object')


count                                                                 \
     PolicyNum                                                                  
Year      2006   2007   2008   2009   2010   2006   2007   2008   2009   2010   
Freq                                                                            
0        845.0  781.0  816.0  811.0  707.0  845.0  781.0  816.0  811.0  707.0   
1        153.0  146.0  155.0  155.0  209.0  153.0  146.0  155.0  155.0  209.0   
2         67.0   92.0   56.0   69.0   86.0   67.0   92.0   56.0   69.0   86.0   
3         40.0   47.0   41.0   26.0   40.0   40.0   47.0   41.0   26.0   40.0   
4         13.0   27.0   24.0   17.0   18.0   13.0   27.0   24.0   17.0   18.0   
5          9.0   15.0    6.0   11.0   12.0    9.0   15.0    6.0   11.0   12.0   
6          6.0    6.0    5.0    1.0    9.0    6.0    6.0    5.0    1.0    9.0   
7          1.0    8.0    5.0    4.0    4.0    1.0    8.0    5.0    4.0    4.0   
8          6.0    4.0    2.0    2.0    6.0    6.0    4.0    2.0    2.0    6.0   
9          3.0    1.0    2.0    2.0    1.0    3.0    1.0    2.0    2.0    1.0   
10         3.0    NaN    3.0    2.0    3.0    3.0    NaN    3.0    2.0    3.0   
11         2.0    NaN    1.0    1.0    2.0    2.0    NaN    1.0    1.0    2.0   
12         NaN    1.0    1.0    1.0    NaN    NaN    1.0    1.0    1.0    NaN   
13         NaN    NaN    1.0    1.0    1.0    NaN    NaN    1.0    1.0    1.0   
14         1.0    1.0    1.0    NaN    2.0    1.0    1.0    1.0    NaN    2.0   
15         NaN    1.0    NaN    2.0    1.0    NaN    1.0    NaN    2.0    1.0   
16         NaN    2.0    NaN    1.0    2.0    NaN    2.0    NaN    1.0    2.0   
17         NaN    1.0    NaN    NaN    1.0    NaN    1.0    NaN    NaN    1.0   
18         NaN    1.0    2.0    1.0    1.0    NaN    1.0    2.0    1.0    1.0   
19         NaN    NaN    NaN    1.0    1.0    NaN    NaN    NaN    1.0    1.0   
20         1.0    NaN    1.0    NaN    NaN    1.0    NaN    1.0    NaN    NaN   
21         NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN   
24         1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN   
26         NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN   
30         NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0   
38         NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN   
39         NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0   
55         NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN   
58         1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN   
67         NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN   
100        1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN   
103        NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0   
143        NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN   
157        NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN   
208        1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN   
212        NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN   
223        NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN    NaN   
228        NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN   
239        NaN    NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0   
263        NaN    NaN    NaN    1.0    NaN    NaN    NaN    NaN    1.0    NaN   

              sum                                                      \
        PolicyNum                                                       
Year         2006         2007         2008         2009         2010   
Freq                                                                    
0     128120929.0  118356957.0  123683769.0  122515551.0  107578625.0   
1      22363064.0   21543573.0   22463355.0   22570967.0   30622820.0   
2     

## Groupby

In [49]:
print(claims.groupby(['Year','Freq'])['PolicyNum'].count())

print(claims.groupby(['Year','Freq'])['PolicyNum'].count().groupby(level='Year').sum())

100*(claims.groupby(['Year','Freq'])['PolicyNum'].count().loc[(2010,slice(None))]/claims.groupby(['Year','Freq'])['PolicyNum'].count().groupby(level='Year').sum()[2010])

Year  Freq
2006  0       845
      1       153
      2        67
      3        40
      4        13
             ... 
2010  19        1
      30        1
      39        1
      103       1
      239       1
Name: PolicyNum, Length: 103, dtype: int64
Year
2006    1154
2007    1138
2008    1125
2009    1112
2010    1110
Name: PolicyNum, dtype: int64


Year  Freq
2010  0       63.693694
      1       18.828829
      2        7.747748
      3        3.603604
      4        1.621622
      5        1.081081
      6        0.810811
      7        0.360360
      8        0.540541
      9        0.090090
      10       0.270270
      11       0.180180
      13       0.090090
      14       0.180180
      15       0.090090
      16       0.180180
      17       0.090090
      18       0.090090
      19       0.090090
      30       0.090090
      39       0.090090
      103      0.090090
      239      0.090090
Name: PolicyNum, dtype: float64

In [60]:
# not working

output = claims.groupby(['Year','Freq'])['PolicyNum'].count().rename('Percentage').transform(lambda x: x/x.sum())
output

Year  Freq
2006  0       0.149849
      1       0.027132
      2       0.011882
      3       0.007093
      4       0.002305
                ...   
2010  19      0.000177
      30      0.000177
      39      0.000177
      103     0.000177
      239     0.000177
Name: Percentage, Length: 103, dtype: float64

## Last Attempt:

To determine the (percentage) proportion of policies for each claim frequency, we will first group the "Year" and "Freq".   To calculate the percentage **within each "Year" group**, the following command can be used groupby(level=0).apply(lambda x: 100*x/x.sum())

**Note:** Because the original dataframe becomes a multiple index dataframe after grouping, the level = 0 refers to the top level index, which in our case is 'Year'.

You can see the results below, which have already been sorted by percent of sales contribution for each sales person.

In [88]:
#Table1 = claims.groupby(['Year','Freq']).agg({'PolicyNum' : 'count'}).groupby(level='Year').apply(lambda x: 100*x/x.sum())

Table1 = claims.groupby(['Year','Freq']).agg({'PolicyNum' : 'count'}).groupby(level='Year').apply(lambda x: 100*x/x.sum())

Table1.rename(columns={'PolicyNum':'Percentage'})

Table1.loc[(2010,slice(None))]


PolicyNum
Freq           
0     63.693694
1     18.828829
2      7.747748
3      3.603604
4      1.621622
5      1.081081
6      0.810811
7      0.360360
8      0.540541
9      0.090090
10     0.270270
11     0.180180
13     0.090090
14     0.180180
15     0.090090
16     0.180180
17     0.090090
18     0.090090
19     0.090090
30     0.090090
39     0.090090
103    0.090090
239    0.090090

2. From those 403 policyholders who made at least one claim, create a table that provides information about the distribution of average claim amounts in year 2010.

2.1. What is the mean of the average claim amounts?

2.2. What is the third quartile of the average claim amounts?

First, we add the column, namely `ClaimsAvg` representing the average cost per claim for each observation. The average cost per claim (or claim average) amount is calculated by dividing the number of claims  by the total claim amount.

In [53]:
claims['ClaimsAvg'] = claims['y']/claims['Freq']
claims['ClaimsAvg'] = claims['ClaimsAvg'].fillna(0)
claims['ClaimsAvg']

0            0.00
1            0.00
2            0.00
3            0.00
4         6838.87
          ...    
5634         0.00
5635    168304.05
5636      1034.33
5637         0.00
5638         0.00
Name: ClaimsAvg, Length: 5639, dtype: float64

The information about the distribution of average claim amounts in year 2010 is given in the table below including count, mean, std, min, 25, 50, 75% percentiles and max.

In [423]:
#claimsFreqDist.query('Year == 2010 & Freq >= 1')



#claims[['Year','y','Freq']][0:10]
#claims[(claims['Year']==2010) &  (claims['Freq'] >= 1)]['y'].describe()
claims[(claims['Year']==2010) &  (claims['Freq'] >= 1)]['ClaimsAvg'].describe()

count    4.030000e+02
mean     5.633195e+04
std      6.507344e+05
min      1.666667e+02
25%      2.225760e+03
50%      4.951000e+03
75%      1.190050e+04
max      1.292222e+07
Name: ClaimsAvg, dtype: float64

2.1. What is the mean of the average claim amounts?

In [57]:
# the mean of the average claim amounts

(claims[(claims['Year']==2010) &  (claims['Freq'] >= 1)]['ClaimsAvg']).mean()

ans2_1 = (claims[(claims['Year']==2010) &  (claims['Freq'] >= 1)]['ClaimsAvg']).mean()

print('Ans: The mean of the average claim amounts: ', ans2_1)

Ans: The mean of the average claim amounts:  56331.94660006498


2.2 What is the third quartile of the average claim amounts?

In [60]:
# the third quartile of the average claim amounts
(claims[(claims['Year']==2010) &  (claims['Freq'] >= 1)]['ClaimsAvg']).quantile(q = 0.75)

ans2_2 = (claims[(claims['Year']==2010) &  (claims['Freq'] >= 1)]['ClaimsAvg']).quantile(q = 0.75)

print('Ans: the third quartile of the average claim amounts:', ans2_2)

Ans: the third quartile of the average claim amounts: 11900.5


In [61]:
# the mean of the average claim amounts

# (claims[(claims['Year']==2010) &  (claims['Freq'] >= 1)]['yAvg']).mean()

In [62]:
# the third quartile of the average claim amounts
# (claims[(claims['Year']==2010) &  (claims['Freq'] >= 1)]['yAvg']).quantile(q = 0.75)

3. Consider the claims data over the 5 years between 2006-2010 inclusive. Create a table that show the average claim varies over time, average frequency, average coverage and the number of policyholders. 

3.1 What can you say about the number of policyholders over this period?

3.2 How does the average coverage change over this period?

In [425]:
claims.columns

Index(['PolicyNum', 'Year', 'LnCoverage', 'BCcov', 'Premium', 'Freq', 'Deduct',
       'y', 'lny', 'yAvg', 'lnDeduct', 'Fire5', 'NoClaimCredit', 'TypeCity',
       'TypeCounty', 'TypeMisc', 'TypeSchool', 'TypeTown', 'TypeVillage',
       'AC00', 'AC05', 'AC10', 'AC15', 'ClaimsAvg'],
      dtype='object')

In [434]:
claims.groupby('Year').agg({'Freq':'mean', 'ClaimsAvg':'mean', 'BCcov':'mean', 'PolicyNum':'count' })

Freq     ClaimsAvg         BCcov  PolicyNum
Year                                                 
2006  0.951473   9694.992379  3.249819e+07       1154
2007  1.166960   6543.947626  3.527595e+07       1138
2008  0.974222   5310.505387  3.726748e+07       1125
2009  1.219424   4571.949986  4.035538e+07       1112
2010  1.240541  20452.049081  4.124207e+07       1110

In [63]:
# claims.groupby('Year').agg({'Freq':'mean', 'yAvg':'mean', 'BCcov':'mean', 'PolicyNum':'count' })

## Conclusion

1. The table shows that the average claim varies over time, especially with the high 2010 value (that we saw was due to a single large claim).

2. The total number of policyholders is steadily declining and, conversely, the coverage is steadily increasing (**Answers of Questions 3.1 and 3.2**).

3. The coverage variable is the amount of coverage of the property and contents. Roughly, you can think of it as the maximum possible payout of the insurer.